TUR-21619: Prototype Custom Scheduler

ciaconet opened the pull request:
### Summary
[TUR-21619](https://ntopology.atlassian.net/browse/TUR-21619) is a prototype method of executing groups of pytest test cases in series with each group running it's specific test cases in parallel with a variable number of workers.

Tests can be marked with a custom mark. This custom mark specifies an arbitrary group name, and the number of processes to run tests in that group with, with the number of processes separated from the group name by an underscore.

Due to limitations in `pytest`/`execnet` a shutdown signal must be sent to a node when one test case remains on the node (the test case is actually ran immediately, but we do not receive feedback through the channel until a shutdown signal is sent due to an old design decision in `pytest`). Because of this, in order to run additional tests after a group is nearly complete (when each worker has at most 1 test remaining), we must shutdown each node, recreate each node, and then schedule more tests. This adds overhead as we must teardown + startup nodes, however, this overhead is on the order of 1 second per teardown/startup. If this custom scheduling method allows us to save more than 1 second in test execution, it is worth the increased overhead cost.

Note: The `-n` argument with `--dist customgroup` is the **maximum** number of worker processes. If `-n` specifies `4` but a specific group specifies `8`, the specific group will be ran across only 4 nodes, we will never spin up more nodes than the `-n` argument allows.

**Prototype, not necessarily meant to be final implementation. Lots of messy comments/not well documented**.

### Testing
- See `README.md` added in this PR
- Test [windows ntop-tests-silver-no-render](http://35.227.27.17:8080/job/turbo-ondemand-all/7/) command line only, skipping build
- Test [linux ntop-tests-silver-no-render](http://35.227.27.17:8080/job/turbo-ondemand-all/6/) skipping build

[TUR-21619]: https://ntopology.atlassian.net/browse/TUR-21619?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ciaconet committed:
changes to add customgroup.py not working currently

ciaconet committed:
add test

ciaconet committed:
kind of working by loading when a node node2pending is 1 instead of 0

ciaconet committed:
kind of working with node teardown + recreate, works with -n 1, needs work with -n more than 1, appears test is marked complete AFTER we do node teardown, we need to make sure we do teardown after all tests are complete + marked as such

ciaconet committed:
prototype approach working fully with -n 4, needs further validation and is messy

ciaconet committed:
remove log statements used to debug

ciaconet committed:
add newline for no overlapping in terminal output, update README for local testing of customgroup

Heroico committed:
TUR-21619
Create only as many workers as necessary

minor additions

Better integration with terminal dist reporter

TUR-21619
delayed shutting down until last test case

ciaconet committed:
gate dsession.py changes on if statements, clean up commented out code/unused variables

ciaconet committed:
remove commented out breakpoint

Heroico commented on src/xdist/dsession.py:
```diff
@@ -198,6 +239,11 @@ def worker_workerfinished(self, node: WorkerController) -> None:
+        if self.remake_nodes:
+            node.ensure_teardown()
+            self._active_nodes.remove(node)
+            self.do_breakpoint = True
```
I think you removed the `do_breakpoint` member variable

ciaconet replied:
good catch thanks!

Heroico commented on src/xdist/dsession.py:
```diff
@@ -147,14 +153,49 @@ def loop_once(self) -> None:
             except Empty:
+                # Custom logic for CustomGroup scheduler:
+                if isinstance(self.sched, CustomGroup):
+                    all_one = True
```
I'd rename this variable if we want to share with the upstream - something like `all_nodes_finishing` or something.

ciaconet replied:
Renamed that, that makes more sense, thanks

Heroico commented on src/xdist/scheduler/customgroup.py:
```diff
@@ -0,0 +1,349 @@
+
+    @property
+    def tests_finished(self) -> bool:
+        """Return True if all tests have been executed by the nodes."""
```
Maybe we should call out the "more than one test remaining" concept.

ciaconet replied:
Just in the docstring?

Heroico reviewed (commented):
Direction looks good.
The `do_breakpoint` variable should be completely removed.

ciaconet committed:
pr feedback

ciaconet committed:
add trylast

ciaconet committed:
small bugfix

ciaconet committed:
allow increased number of workers

Heroico commented on src/xdist/workermanage.py:
```diff
@@ -82,9 +82,12 @@ def rsync_roots(self, gateway: execnet.Gateway) -> None:
     def setup_nodes(
         self,
         putevent: Callable[[tuple[str, dict[str, Any]]], None],
+        max_nodes: Union[int, None] = None    
```
Why not `Optional`?

ciaconet replied:
Changed thanks

ciaconet committed:
typing change

ciaconet committed:
clean up README.md and other small cleanup

Heroico commented on src/xdist/scheduler/customgroup.py:
```diff
@@ -0,0 +1,348 @@
+                if self.pending_groups:
+                    dist_group_key = self.pending_groups.pop(0)
+                    dist_group = self.dist_groups[dist_group_key]
+                    nodes = cycle(self.nodes[0:dist_group['group_workers']])
```
I'm still confused by this:
- `check_schedule` takes a node as input. it seems to be oriented to "maybe scheduling on a single node"
- this code here assigns to any nodes in the dist group
The "reschedule" calls this method with the "first node" which is not guaranteed to have anything assigned to it.
Any thoughts?

ciaconet replied:
I left the check schedule signature unchanged, but we don't do anything if any node is working, so it is essentially node agnostic

Heroico replied:
I apologize but I don't understand. `check_schedule` returns early if the node is checking down.
So there's the possibility that `node[0]` is shutting down while others are running (or viceversa) - i.e. this cares about which node is called on.

ciaconet replied:
Yes but if node[0] is shutting down we won't be scheduling anyways. We can remove that check though, I just didn't want to change the method signature, but we could also change and not pass in a node at all

Heroico replied:
I'm more worried about the opposite: `node[0]` is not shutting down, so it jumpts to more scheduling, while the others are still running tests.
Also, since regular `schedule` is supposed to be `check_schedule` for all nodes after its first call, might it not be clearer to use that one?

ciaconet replied:
We check that they are all idle, so we wouldn't be rescheduling. We can remove the usage of node entirely.

I'm unsure what you mean, I think check_schedule is more clear rather than continuously calling schedule, but if behavior is the same either way I don't think it matters

Heroico commented:
Identified one race condition: when the following conditions are met:
- tests crash
-  some of those add skip dynamically (e.g. [this one](https://github.com/nTopology/turbo/blob/a020fe7526707e3c07ab15f67b2cec59f4c392c3/integration_tests/ntop_tests/functional_tests/transient_thermal_analysis/transient_thermal_analysis_test.py#L83))
then rescheduling might happen at a moment when tests should have been marked as skipped but they crashed instead so "pending" is inaccurate: if so:
- hangs up or crashes with an internal error during scheduling

One manifestation was when an entire distgroup in the scheduler were crashed/skipped tests.

The solution is implemented the last [commit](https://github.com/nTopology/pytest-xdist/pull/2/commits/3470d353dee44f1889b1a1cab63486ce90fa71d8) of this [PR](https://github.com/nTopology/pytest-xdist/pull/2) - it avoids rescheduling at all if many failures are observed, and if rescheduling fails, it gracefully shuts down so an xml is generated.

I'm leaning towards including my PR as part of the prototype before rolling it out.

Heroico committed:
TUR-21619: Integrated CustomGroup into DSession event handling to avoid multiple node teardown/setyup between marks. (#2)

* TUR-21619
mini-restructuring with logging to investigate concurrency/scheduling.

* TUR-21619
Integrated CustomGroup into DSession event handlers.

* TUR-21619
Improved custom group logging to be flagged by verbosity.
Handled edge case of empty collection for a mark.

* TUR-21619
Identified one race condition tied to  DSession 'shouldstop' -
since it should stop, there's little point in rescheduling tests if it's supposed to wind down.

* TUR-21619
Code cleanu - removed logging and unused code.

Heroico commented:
The latest commit addresses the race conditions I observed.
